File: android/src/main/java/org/maplibre/reactnative/components/annotation/MLRNMarkerViewManager.java
package org.maplibre.reactnative.components.annotation;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;
import org.maplibre.reactnative.components.AbstractEventEmitter;
import org.maplibre.reactnative.utils.GeoJSONUtils;

import java.util.Map;

public class MLRNMarkerViewManager extends AbstractEventEmitter<MLRNMarkerView> {
    public static final String REACT_CLASS = "MLRNMarkerView";

    public MLRNMarkerViewManager(ReactApplicationContext reactApplicationContext) {
        super(reactApplicationContext);
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @ReactProp(name="coordinate")
    public void setCoordinate(MLRNMarkerView markerView, String geoJSONStr) {
        markerView.setCoordinate(GeoJSONUtils.toPointGeometry(geoJSONStr));
    }

    @ReactProp(name="anchor")
    public void setAnchor(MLRNMarkerView markerView, ReadableMap map) {
        markerView.setAnchor((float) map.getDouble("x"), (float) map.getDouble("y"));
    }

    @Override
    protected MLRNMarkerView createViewInstance(ThemedReactContext reactContext) {
        return new MLRNMarkerView(reactContext, this);
    }

    @Override
    public Map<String, String> customEvents() {
        return MapBuilder.<String, String>builder()
                .build();
    }
}

File: android/src/main/java/org/maplibre/reactnative/components/annotations/MLRNMarkerViewManager.kt
package org.maplibre.reactnative.components.annotations

import android.view.View
import com.facebook.react.bridge.Dynamic
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.MLRNMarkerViewManagerDelegate
import com.facebook.react.viewmanagers.MLRNMarkerViewManagerInterface

@ReactModule(name = MLRNMarkerViewManager.REACT_CLASS)
class MLRNMarkerViewManager(private val reactApplicationContext: ReactApplicationContext) :
    ViewGroupManager<MLRNMarkerView>(),
    MLRNMarkerViewManagerInterface<MLRNMarkerView> {

    private val delegate: MLRNMarkerViewManagerDelegate<MLRNMarkerView, MLRNMarkerViewManager> =
        MLRNMarkerViewManagerDelegate(this)

    override fun getDelegate(): ViewManagerDelegate<MLRNMarkerView> = delegate

    companion object {
        const val REACT_CLASS: String = "MLRNMarkerView"
    }

    override fun getName(): String = REACT_CLASS

    override fun createViewInstance(reactContext: ThemedReactContext): MLRNMarkerView {
        return MLRNMarkerView(reactContext)
    }

    override fun addView(parent: MLRNMarkerView, child: View, index: Int) {
        parent.addView(child, index)
    }

    override fun removeViewAt(parent: MLRNMarkerView, index: Int) {
        parent.removeViewAt(index)
    }

    override fun getChildCount(parent: MLRNMarkerView): Int {
        return parent.childCount
    }

    override fun getChildAt(parent: MLRNMarkerView, index: Int): View? {
        return parent.getChildAt(index)
    }

    @ReactProp(name = "lngLat")
    override fun setLngLat(markerView: MLRNMarkerView, lngLat: Dynamic) {
        if (lngLat.type == ReadableType.Array) {
            val arr = lngLat.asArray()
            if (arr != null && arr.size() >= 2) {
                markerView.setLngLat(doubleArrayOf(arr.getDouble(0), arr.getDouble(1)))
            }
        }
    }

    @ReactProp(name = "anchor")
    override fun setAnchor(markerView: MLRNMarkerView, map: ReadableMap?) {
        if (map != null) {
            markerView.setAnchor(map.getDouble("x").toFloat(), map.getDouble("y").toFloat())
        }
    }

    @ReactProp(name = "allowOverlap")
    override fun setAllowOverlap(markerView: MLRNMarkerView, allowOverlap: Boolean) {
        markerView.setAllowOverlap(allowOverlap)
    }

    @ReactProp(name = "isSelected")
    override fun setIsSelected(markerView: MLRNMarkerView, isSelected: Boolean) {
        markerView.setIsSelected(isSelected)
    }
}
